### PR TITLE
Update audio guidance for AA compatibility

### PR DESCRIPTION
**Pull request by kmadsen**
### Description

Fixes NAVAND-628

AA and the nav-sdk work on different release channels. In order to support long term backwards compatibility, AA will only use public apis from the nav-sdk. 

`MapboxAudioGuidance` is an internal api, so this moves it out into the public. Looking to make this part of the 2.9 release train, and the alpha release starts today.

Some related issues and pull requests that this change is addressing
 - https://github.com/mapbox/mapbox-navigation-android/pull/6316
 - https://github.com/mapbox/mapbox-navigation-android/pull/6320
 - https://github.com/mapbox/mapbox-navigation-android/pull/6304
 - https://github.com/mapbox/mapbox-navigation-android/pull/6305
 - https://github.com/mapbox/mapbox-navigation-android/pull/6303

## Code samples

Here are some samples of the interactions.

``` kotlin
// If MapboxAudioGuidance is not available, create it and register it. This allows for
// different systems to share an instance of MapboxAudioGuidance
val audioGuidance = MapboxAudioGuidance.getInstance()
```

Lower level interactions are available, but you need to know what you're doing.
``` kotlin
// 1. Get and/or enable audio guidance
if (MapboxNavigationApp.getObservers(MapboxAudioGuidance::class).isEmpty()) {. 
  MapboxNavigationApp.registerObserver(MapboxAudioGuidance.create())
}

// 2. Use the audio guidance service -- assumes it has been registered or else it will crash
MapboxNavigationApp.getObserver(MapboxAudioGuidance::class).mute()

// 3. Remove it - code like 1 may re-register. code like 2 will crash if there are components that assume it's registered
MapboxNavigationApp.getObservers(MapboxAudioGuidance::class)?.firstOrNull()?.let {
  MapboxNavigationApp.unregisterObserver(it)
}
```

### CODE REVIEW

**Comment by kmadsen** [comment]
Running into an issue from this approach https://github.com/mapbox/mapbox-navigation-android/pull/6316

It works fine while debugging but the incompatibility is still caught so there are no temporary merges

```
./gradlew libnavui-androidauto:jacocoTestReport
```
```
> Task :libnavui-androidauto:compileReleaseKotlin FAILED
e: .../mapbox/androidauto/MapboxCarApp.kt: (7, 43): Unresolved reference: MapboxAudioGuidance
e: .../mapbox/androidauto/MapboxCarApp.kt: (32, 39): Unresolved reference: MapboxAudioGuidance
e: .../mapbox/androidauto/MapboxCarApp.kt: (33, 29): Not enough information to infer type variable T
e: .../mapbox/androidauto/MapboxCarApp.kt: (33, 41): Unresolved reference: MapboxAudioGuidance
e: .../mapbox/androidauto/MapboxCarApp.kt: (47, 33): Not enough information to infer type variable T
e: .../mapbox/androidauto/MapboxCarApp.kt: (47, 46): Unresolved reference: MapboxAudioGuidance
e: .../mapbox/androidauto/MapboxCarApp.kt: (48, 50): Unresolved reference: MapboxAudioGuidance
e: .../mapbox/androidauto/navigation/audioguidance/AppAudioGuidanceUi.kt: (11, 43): Unresolved reference: MapboxAudioGuidanceState
e: .../mapbox/androidauto/navigation/audioguidance/AppAudioGuidanceUi.kt: (43, 36): Unresolved reference: MapboxAudioGuidanceState
e: .../mapbox/androidauto/navigation/audioguidance/AppAudioGuidanceUi.kt: (45, 13): Not enough information to infer type variable R
e: .../mapbox/androidauto/navigation/audioguidance/AppAudioGuidanceUi.kt: (46, 17): Variable expected
```

**Comment by kmadsen** [comment]
Converting to a draft because we need to release a new version of andriodauto before we can proceed here

**Comment by kmadsen** [comment]
> While we are on the topic of the MapboxAudioGuidance interface, what do you think about simplifying it to just?

Removing the interface in favor of a class. Which is more consistent with the rest of the sdk. As for the change from `fun setMuted(muted: Boolean)` vs `mute` and `unmute`

These definitions were built to be consistent with other existing stable apis
https://github.com/mapbox/mapbox-navigation-android/blob/8df22ed208982a52fe1912a4522f4862cc8e2d74/libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/view/MapboxAudioGuidanceButton.kt#L117-L130

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt [hunk 1]
```diff
@@ -35,30 +33,41 @@ import kotlinx.coroutines.launch
 /**
  * Implementation of [MapboxAudioGuidance]. See interface for details.
  */
-@ExperimentalPreviewMapboxNavigationAPI
-class MapboxAudioGuidanceImpl(
+class MapboxAudioGuidance
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt:36]: There's a few issues with the interface, would you like to discuss it?

Also, we do not have interfaces anywhere in public apis so this pull request is attempting to introduce solutions that are consistent

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt [hunk 2]
```diff
@@ -35,30 +33,41 @@ import kotlinx.coroutines.launch
...
+internal constructor(
     private val audioGuidanceServices: MapboxAudioGuidanceServices,
-    private val configOwner: NavigationConfigOwner,
-    dispatcher: CoroutineDispatcher = Dispatchers.Main,
-) : MapboxAudioGuidance {
+    dispatcher: CoroutineDispatcher,
+) : MapboxNavigationObserver {
 
-    var dataStoreOwner: NavigationDataStoreOwner? = null
+    constructor() : this(MapboxAudioGuidanceServices(), Dispatchers.Main)
```
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt:43]: that's an idea but i like the empty constructor with a constructor that is visible for testing

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt:36]: Discussed offline some and to continue to provide the ability to override the behavior will make `open`

File: libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (kmadsen) [reply libnavui-voice/src/main/java/com/mapbox/navigation/ui/voice/api/MapboxAudioGuidance.kt]: @tomaszrybakiewicz how about this, added a couple companion functions to make to make it clear what the intended usage is.

``` kotlin
MapboxAudioGuidance {
..
companion object {
    fun create(): MapboxAudioGuidance ...

    fun getInstance(): MapboxAudioGuidance = MapboxNavigationApp
            .getObservers(MapboxAudioGuidance::class)
            .firstOrNull() ?: MapboxAudioGuidance(MapboxAudioGuidanceServices(), Dispatchers.Main)
            .also { MapboxNavigationApp.registerObserver(it) }
```

**Review by tomaszrybakiewicz — APPROVED** [review]

**Comment by kmadsen** [comment]
> When running this branch and starting MapboxNavigationViewCustomizedActivity, the QA-Test-App is crashing for me

Thanks for catching! I didn't fix AudioGuidanceStateController

You can still use `MapboxNavigationApp.getObserver(MapboxAudioGuidance::class)`

Using `MapboxAudioGuidance.getInstance()` will ensure MapboxAudioGuidance is available

**Comment by kmadsen** [comment]
> Out of curiosity, we are introducing MapboxAudioGuidanceState in libnavui-voice, but we also have AudioGuidanceState in libnavui-app. Why do we have 2 states for audio guidance?

I'm not seeing a need for the `AudioGuidanceState` either.

**Pull request merged**